File: seapath_test/test_retrieve_gnss.cpp
#include <iostream>
#include <vector>
#include <string>
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <unistd.h>
#include <ctime>
#include <sstream>
#include <algorithm>

const double pi = 3.14159265358979323846;

double ORIGIN_N = 0.0;
double ORIGIN_E = 0.0;

struct SeapathData {
    double north = 0.0;
    double north_sigma = 0.0;

    double east = 0.0;
    double east_sigma = 0.0;

    double altitude = 0.0;
    double altitude_sigma = 0.0;

    double heading = 0.0;

    double sog_mps = 0.0;
    double cog_rad = 0.0;

    double rot = 0.0;
};

std::pair<double, double> displacement_wgs84(double north, double east) {
    double R = 6371.0;
    double m_per_deg_lat = R * 1000 * (pi / 180.0);
    double m_per_deg_lon = m_per_deg_lat * cos(ORIGIN_N * pi / 180.0);

    double displacement_north = (north - ORIGIN_N) * m_per_deg_lat;
    double displacement_east = (east - ORIGIN_E) * m_per_deg_lon;

    return {displacement_north, displacement_east};
}

double convert_dms_to_dd(double dms) {
    double degrees = static_cast<int>(dms / 100);
    double minutes = dms - (degrees * 100);
    double dd = degrees + (minutes / 60);
    return round(dd * 1e6) / 1e6;
}

SeapathData parse_nmea_data(std::string nmea_data) {
    SeapathData seapath_data;
    std::istringstream dataStream(nmea_data);
    std::string sentence;

    while (getline(dataStream, sentence, '\n')) {
        std::string data = sentence.substr(1, sentence.find('*') - 1);
        std::vector<std::string> parts;
        std::string part;
        std::istringstream partStream(data);
        while (getline(partStream, part, ',')) {
            parts.push_back(part);
        }

        if (parts[0] == "INGGA") { // GNSS fix

            if(parts[2].empty() || parts[4].empty() || parts[9].empty()) {
                continue;
            }

            std::string north = parts[2];
            std::string east = parts[4];
            std::string altitude = parts[9];


            seapath_data.north = convert_dms_to_dd(std::stod(north));
            seapath_data.east = convert_dms_to_dd(std::stod(east));
            seapath_data.altitude = std::stod(altitude);

            if (ORIGIN_N == 0.0) {
                ORIGIN_N = seapath_data.north;
                ORIGIN_E = seapath_data.east;
            }

            auto [n, e] = displacement_wgs84(seapath_data.north, seapath_data.east);
            std::cout << "N [m]: " << n << " \t E [m]: " << e << std::endl;
        }

        if (parts[0] == "INHDT") { // True heading
            if(parts[1].empty()) {
                continue;
            }

            std::string true_north_heading = parts[1];

            seapath_data.heading = std::stod(true_north_heading);
        }

        if (parts[0] == "INVTG") { // Course over ground and speed over ground
            if(parts[1].empty() || parts[7].empty()) {
                continue;
            }
            std::string ground_speed_mps = parts[7];
            std::string ground_course_rad = parts[1];

            seapath_data.sog_mps = std::round(std::stod(ground_speed_mps) * 1000.0 / 3600.0 * 1e6) / 1e6;
            seapath_data.cog_rad = std::round(std::stod(ground_course_rad) * 2 * pi / 360.0 * 1e6) / 1e6;
        }

        if (parts[0] == "GNGST") { // GNSS pseudorange error statistics
            if (parts[6].empty() || parts[7].empty() || parts[8].empty()) {
                continue;
            }

            std::string sigma_lat = parts[6];
            std::string sigma_lon = parts[7];
            std::string sigma_alt = parts[8];

            seapath_data.north_sigma = std::stod(sigma_lat);
            seapath_data.east_sigma = std::stod(sigma_lon);
            seapath_data.altitude_sigma = std::stod(sigma_alt);
        }

        if (parts[0] == "INROT") { // Rate of turn
            if(parts[1].empty()) {
                continue;
            }
            std::string rate_of_turn = parts[1];

            seapath_data.rot = std::stod(rate_of_turn);
        }
    }

    return seapath_data;
}


int main() {
    const char *UDP_IP = "0.0.0.0";
    const int UDP_PORT = 31420;

    int sockfd;
    char buffer[1024];
    struct sockaddr_in servaddr, cliaddr;

    if ((sockfd = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
        perror("socket creation failed");
        exit(EXIT_FAILURE);
    }

    memset(&servaddr, 0, sizeof(servaddr));
    memset(&cliaddr, 0, sizeof(cliaddr));

    servaddr.sin_family = AF_INET;
    servaddr.sin_addr.s_addr = INADDR_ANY;
    servaddr.sin_port = htons(UDP_PORT);

    if (bind(sockfd, (const struct sockaddr *)&servaddr, sizeof(servaddr)) < 0) {
        perror("bind failed");
        exit(EXIT_FAILURE);
    }

    while(true) {
        int len = sizeof(cliaddr);
        int n = recvfrom(sockfd, (char *)buffer, sizeof(buffer), MSG_WAITALL, (struct sockaddr *) &cliaddr, (socklen_t *)&len);
        buffer[n] = '\0';

        SeapathData seapath_data = parse_nmea_data(buffer);
        //std::cout << seapath_data.north << ", " << seapath_data.east << ", " << seapath_data.altitude << std::endl;
    }

    return 0;
}

File: seapath_test/test_retrieve_gnss.py
#!/usr/bin/python3
import socket

import numpy as np

from dataclasses import dataclass

ORIGIN_N = None
ORIGIN_E = None

@dataclass
class SeapathData:
    north: float = None
    north_sigma: float = None

    east: float = None
    east_sigma: float = None

    altitude: float = None
    altitude_sigma: float = None

    heading: float = None

    sog_mps: float = None
    cog_rad: float = None

    rot: float = None

def displacement_wgs84(north, east):
    # Approximate radius of earth in km, at around 45 degrees latitude
    R = 6371.0

    # Distance per degree latitude in meters
    m_per_deg_lat = R * 1000 * (np.pi / 180.0)

    # Distance per degree longitude at this latitude in meters
    m_per_deg_lon = m_per_deg_lat * np.cos(np.radians(ORIGIN_N))

    # Displacement calculation
    displacement_north = (north - ORIGIN_N) * m_per_deg_lat
    displacement_east = (east - ORIGIN_E) * m_per_deg_lon

    return displacement_north, displacement_east

def convert_dms_to_dd(dms):
    degrees = int(dms / 100)
    minutes = dms - (degrees * 100)
    dd = degrees + (minutes / 60)
    return round(dd, 6)  # limiting the precision to 6 decimal places


def parse_nmea_data(nmea_data):

    global ORIGIN_N, ORIGIN_E

    seapath_data = SeapathData()

    sentences = nmea_data.decode().strip().split('\n')

    # Process each sentence
    for sentence in sentences:

        data, checksum = sentence[1:].split('*')

        parts = data.split(',')
        # Check the sentence type and print the relevant parts
        if parts[0] == 'INGGA': # GNSS fix
            north = parts[2]
            east = parts[4]
            altitude = parts[9]

            if north == "" or east == "":
                continue

            seapath_data.north = convert_dms_to_dd(float(north))
            seapath_data.east = convert_dms_to_dd(float(east))
            seapath_data.altitude = float(altitude)

            if ORIGIN_N is None:
                ORIGIN_N = seapath_data.north
                ORIGIN_E = seapath_data.east

            n, e = displacement_wgs84(seapath_data.north, seapath_data.east)
            #print(f"N [m]: {n} \t E [m]: {e}")

        if parts[0] == 'INHDT': # True heading
            true_north_heading = parts[1]
            if true_north_heading == "":
                continue

            seapath_data.heading = float(true_north_heading)

        if parts[0] == "INVTG": # Course over ground and speed over ground


            ground_speed_mps = parts[7]
            ground_course_rad = parts[1]

            if ground_speed_mps == "" or ground_course_rad == "":
                continue

            seapath_data.sog_mps = np.round(float(ground_speed_mps)*1000.0/3600.0, 6)
            seapath_data.cog_rad = np.round(float(ground_course_rad)* 2 * np.pi / 360.0, 6)


        if parts[0] == "GNGST": #GNSS pseudorange error statistics
            utc_time = parts[1] 
            sigma_fix = parts[2] 
            sigma_semimajor = parts[3] # Standard deviation of semi-minor axis of error ellipse
            sigma_semiminor = parts[4] # Standard deviation of semi-major axis of error ellipse
            error_ellipse_orientation = parts[5] # True north degrees

            sigma_lat = parts[6] 
            sigma_lon = parts[7] 
            sigma_alt = parts[8] 

            if sigma_lat == "" or sigma_lon == "" or sigma_alt == "":
                continue

            seapath_data.north_sigma = float(sigma_lat)
            seapath_data.east_sigma = float(sigma_lon)
            seapath_data.altitude_sigma = float(sigma_alt)

        if parts[0] == "INROT": # Rate of turn
            rate_of_turn = parts[1]

            if rate_of_turn == "":
                continue
            
            seapath_data.rot = rate_of_turn
            

    return seapath_data

UDP_IP = "0.0.0.0"
UDP_PORT = 31420

sock = socket.socket(socket.AF_INET, # Internet
                     socket.SOCK_DGRAM) # UDP
print(f"Binding to {UDP_IP}:{UDP_PORT}...")
sock.bind((UDP_IP, UDP_PORT))
print("Done!")

while True:
    data, addr = sock.recvfrom(1024)
    seapath_data = parse_nmea_data(data)
    print(seapath_data)